src/app/model/ChessEngine.ts

```diff
@@ -68,7 +70,70 @@ export class ChessEngine implements ChessModel {
     return side === Side.WHITE ? Side.BLACK : Side.WHITE;
   }
 
+  willBeKingChecked(squareFrom: Square, squareTo: Square): boolean {
+    let isCheckedSquareFlag = false;
+
+    const chessboard: Chessboard = this.board;
+    const proposedChessboard = _.cloneDeep(chessboard);
+    const { squaresWithPiece: proposedSquaresWithPieces } = proposedChessboard;
+
+    const movedPiece = chessboard.onPositionPiece(squareFrom);
+    const positionFrom = `${squareFrom.column}${squareFrom.row}`;
+    const positionTo = `${squareTo.column}${squareTo.row}`;
+    delete proposedSquaresWithPieces[positionFrom];
+    proposedSquaresWithPieces[positionTo] = movedPiece as Piece;
+
+    const allyKingPositionKey = Object.keys(proposedSquaresWithPieces).find((key) => {
+      const isKingName = proposedSquaresWithPieces[key].name === 'King';
+      const isPlayerSide = proposedSquaresWithPieces[key].side === this.currentSide;
+      return isKingName && isPlayerSide;
+    });
+
+    if (!allyKingPositionKey) return (isCheckedSquareFlag = false);
+
+    const kingPosition = {
+      column: allyKingPositionKey[0],
+      row: Number(allyKingPositionKey[1]) as Row,
+    };
+
+    Object.keys(proposedSquaresWithPieces).forEach((key) => {
+      const mappedPiece = proposedSquaresWithPieces[key];
+      const isEnemySide = mappedPiece.side !== this.currentSide;
+
+      if (isEnemySide) {
+        const mappedPiecePosition = {
+          column: key[0],
+          row: Number(key[1]) as Row,
+        };
+        const possibleMappedPieceMoves = mappedPiece.possibleMoves(mappedPiecePosition, proposedChessboard);
+        const isKingPositionOnPossibleEnemyPieceMoves = possibleMappedPieceMoves.some(
+          (checkedPosition) => JSON.stringify(checkedPosition) == JSON.stringify(kingPosition),
+        );
+
+        if (isKingPositionOnPossibleEnemyPieceMoves) isCheckedSquareFlag = isKingPositionOnPossibleEnemyPieceMoves;
+      }
+    });
+    return isCheckedSquareFlag;
+  }
+
+  returnPlayerMovesWithoutThoseThatCauseHisKingToCheck(position: Square): Square[] {
+    const initialPossibleMoves = this.board.onPositionPiece(position)?.possibleMoves(position, this.board) ?? [];
+
+    const filteringFunction = (onePossibleMove: Square) => {
+      const willBeKingChecked = this.willBeKingChecked(position, onePossibleMove);
+      return !willBeKingChecked;
+    };
+
+    const filteredPossibleMoves = initialPossibleMoves.filter(filteringFunction);
+    return filteredPossibleMoves;
+  }
+
   possibleMoves(position: Square): Square[] {
-    return this.board.onPositionPiece(position)?.possibleMoves(position, this.board) ?? [];
+    return this.returnPlayerMovesWithoutThoseThatCauseHisKingToCheck(position);
   }
 }
+
+// player: Player,
+//     chessboard:Chessboard,
+//     squareFrom:Square,
+//     squareTo:Square
```

Comment:
jak usunąłeś player'a to ten deadcode chyba nie jest już potrzebny

Reply:
Tak - napisałem o tym na konwersacji grupowej ;)